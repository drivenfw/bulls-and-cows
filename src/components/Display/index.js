import React, { Component } from 'react'
import styled, { css } from 'styled-components'


const StyledDisplay = styled.div`
  height: 150px;
  padding: .52em .78em;
  border-radius: 10px;
  font-size: 1.25em;
  background: white;

  ${({ theme }) => css`
    border: 1px solid ${theme.primaryColor};
    box-shadow: inset 0 0 3px 2px ${theme.primaryColor2};
    color: ${theme.primaryColor};
  `}
`

const Inner = styled.div`
  width: 86%;
  height: 100%;
  padding-right: 3%;
  overflow-y: scroll;
  line-height: 1.5;
  display: inline-block;
  vertical-align: top;
`

const ScrollBar = styled.div`
  width: 10%;
  height: 100%;
  display: inline-block;
  vertical-align: top;

  & > div:first-child {
    height: 100%; 
    display: flex;
    flex-direction: column;
    justify-content: space-around;
  }
`

const Button = styled.div`
  text-align: right;
  transform: scaleX(1.3) scaleY(3);
  cursor: pointer;
  user-select: none;
  transition: all 0.1s ease;
`

const UpButton = styled(Button)`
  text-shadow: 0 2px 5px ${props => props.theme.primaryColor1};

  &:active {
    transform: scaleX(1.4) scaleY(3) translateY(1px);
    text-shadow: 0 1px 3px ${props => props.theme.primaryColor2};
  }
`

const DownButton = styled(Button)`
  text-shadow: 0 -2px 5px ${props => props.theme.primaryColor1};

  &:active {
    transform: scaleX(1.4) scaleY(3) translateY(-1px);
    text-shadow: 0 -1px 3px ${props => props.theme.primaryColor2};
  }
`

class Display extends Component {
  constructor(props) {
    super(props)
    this.innerEl = React.createRef()
  }

  getLineHeight = () =>
    // TODO: if (window.getComputedStyle) else currentStyle - ?
    getComputedStyle(this.innerEl.current).getPropertyValue('line-height')

  scrollUp = () => {
    this.innerEl.current.scrollTop -= parseFloat(this.getLineHeight())
  } 

  scrollDown = () => {
    this.innerEl.current.scrollTop += parseFloat(this.getLineHeight())
  }

  componentDidMount() {
    window.innerEl = this.innerEl.current
  }

  render() {
    const { children, className } = this.props

    return (
      <StyledDisplay className={className}>
        <Inner innerRef={this.innerEl}>
          {children}
        </Inner>
        <ScrollBar>
          <div>
            <UpButton onClick={this.scrollUp}>⬆</UpButton>
            <DownButton onClick={this.scrollDown}>⬇</DownButton>
          </div>
        </ScrollBar>
      </StyledDisplay>
    )
  }
}

export default Display

