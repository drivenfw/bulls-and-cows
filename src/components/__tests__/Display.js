import React from 'react'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import 'jest-styled-components'

import Display from 'components/Display'


configure({ adapter: new Adapter() })

describe('<Display />', () => {
  describe('snapshot test', () => {
    it('should match the snapshot', () => {
      const tree = mount(<Display />)

      expect(toJson(tree)).toMatchSnapshot()
    })
  })
})

