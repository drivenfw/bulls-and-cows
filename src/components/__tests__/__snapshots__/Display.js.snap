// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Display /> snapshot test should match the snapshot 1`] = `
.c0 {
  height: 150px;
  padding: .52em .78em;
  border-radius: 10px;
  font-size: 1.25em;
  background: white;
  border: 1px solid;
  box-shadow: inset 0 0 3px 2px;
}

.c1 {
  width: 86%;
  height: 100%;
  padding-right: 3%;
  overflow-y: scroll;
  line-height: 1.5;
  display: inline-block;
  vertical-align: top;
}

.c2 {
  width: 10%;
  height: 100%;
  display: inline-block;
  vertical-align: top;
}

.c2 > div:first-child {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c4 {
  text-align: right;
  -webkit-transform: scaleX(1.3) scaleY(3);
  -ms-transform: scaleX(1.3) scaleY(3);
  transform: scaleX(1.3) scaleY(3);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
}

.c3 {
  text-shadow: 0 2px 5px;
}

.c3:active {
  -webkit-transform: scaleX(1.4) scaleY(3) translateY(1px);
  -ms-transform: scaleX(1.4) scaleY(3) translateY(1px);
  transform: scaleX(1.4) scaleY(3) translateY(1px);
  text-shadow: 0 1px 3px;
}

.c5 {
  text-shadow: 0 -2px 5px;
}

.c5:active {
  -webkit-transform: scaleX(1.4) scaleY(3) translateY(-1px);
  -ms-transform: scaleX(1.4) scaleY(3) translateY(-1px);
  transform: scaleX(1.4) scaleY(3) translateY(-1px);
  text-shadow: 0 -1px 3px;
}

@media (min-width:550px) {
  .c0 {
    padding: .52em;
  }
}

@media (min-width:850px) {
  .c0 {
    padding: .52em .78em;
  }
}

<Display>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div
        innerRef={
          Object {
            "current": .c0 {
            width: 86%;
            height: 100%;
            padding-right: 3%;
            overflow-y: scroll;
            line-height: 1.5;
            display: inline-block;
            vertical-align: top;
          }
          
          <div
              class="c0"
          />,
          }
        }
      >
        <div
          className="c1"
        />
      </styled.div>
      <styled.div>
        <div
          className="c2"
        >
          <div>
            <Styled(styled.div)
              onClick={[Function]}
            >
              <styled.div
                className="c3"
                onClick={[Function]}
              >
                <div
                  className="c3 c4"
                  onClick={[Function]}
                >
                  ⬆
                </div>
              </styled.div>
            </Styled(styled.div)>
            <Styled(styled.div)
              onClick={[Function]}
            >
              <styled.div
                className="c5"
                onClick={[Function]}
              >
                <div
                  className="c5 c4"
                  onClick={[Function]}
                >
                  ⬇
                </div>
              </styled.div>
            </Styled(styled.div)>
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</Display>
`;
